Add RestrictNetworkInterfaces= option

mauriciovasquezbernal opened the pull request:
Add `RestrictNetworkInterfaces=` option to limit the network interfaces a program in a unit can use. 

TODO:
- [ ] Investigate compatibility with bpf-firewall as both of them use the same kind and attach type for the BPF programs. In order to use options provided by the bp-firewall together with `RestrictNetworkIntferfaces` both programs should use `BPF_F_ALLOW_MULTI` but it's not used by bpf-firewall: https://github.com/systemd/systemd/commit/acf7f253dedbca7ba5edada37d11221a47101ee3. 
- [ ] Rebase to main (blocked on the outcome of https://github.com/systemd/systemd/pull/1765)
- [ ] How to remove elements from a set of strings?
- [ ] When to detach the eBPF programs https://github.com/systemd/systemd/pull/17655/files#r561982214?
- [ ] Create & configure veth pairs from C in the tests?

Commit:
bpf: add bpf_object helpers for source compiled bpf

Introduce minimalistic set of helpers for bpf programs compiled from
restricted C sources.

Introduce a basic type `struct BPFProgramV2` with 'fd'
and 'attach_type' fields to represent a loaded bpf prog:wqram.
The BPFProgram struct is not used since:
- v2 methods will use libbpf while v1 use raw syscalls
- the majority of its fields is not needed to support BPF program
compiled from sources
- lack of 'attach_type' field

Introduce bpf_object_{} helpers to load bpf programs into kernel, resize
and populate bpf maps, attach program to cgroup hooks.

libbpf dependency must be satisfied to compile the code.

Commit:
bpf: add build script for bpf programs

Add a build script to compile bpf source code. A program in restricted
C is compiled into a temporary object file.
Script generates a single C header file with defined
`const unsigned char[]` array (hexdump) representing object file: each
octet of the object file is an element of the array.
To generate C array script reads *.o file in byte mode in chunks of 16
bytes. Each byte is represented as hexadecimal number with 0x prefix
(hence hexdump), a chunk of 16 numbers forms a line of the target file,
e.g:
```
0x7f, 0x45, 0x4c, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00,
0x00, 0x00, 0x00, 0x00
```

Why hexdump?
- Ready in compile time. No need to distribute *.o files along with
systemd package, the bytecode of BPF program ships with `#include`.
- Transparent and stdout friendly.

In user space hexdump may be included as a regular C header and
read into `struct bpf_object [0] with `bpf_object__open_mem` helper [1].

If build with custom meson build rule, the target header will reside in
build/ directory (not in source tree), e.g the path for allow_bind:
`build/src/core/bpf/allow_bind/allow-bind-hexdump.h`

Summing up, script runs three phases:
* clang to generate LLVM *.bc from restricted C
* llc to compile *.o from *.bc input
* hexdump generation
These phases are logged to stderr for debug purposes.

To include BTF debug information, -g option is passed to clang.

Sample run for `src/core/bpf/allow_bind` program
```
./tools/build-bpf.py --llc_exec /usr/bin/llc --clang_exec /usr/bin/clang
src/core/bpf/allow_bind/allow-bind.c
src/core/bpf/allow_bind/allow-bind-hexdump.h --bpf_hexdump_buffer
allow_bind_hexdump_buffer
DEBUG:root:Generating LLVM bitcode *.bc:
DEBUG:root:/usr/bin/clang -Wno-compare-distinct-pointer-types -O2
-target bpf -emit-llvm -c -D__x86_64__ -I/usr/include/x86_64-linux-gnu/
-I/usr/local/include -I/usr/include -I.
src/core/bpf/allow_bind/allow-bind.c -o /tmp/tmpk8sgch4l.bc
DEBUG:root:Compiling BPF object file:
DEBUG:root:/usr/bin/llc -march=bpf -filetype=obj -o /tmp/tmpk6fxj7j_.o
/tmp/tmpk8sgch4l.bc
DEBUG:root:Generating hexdump for src/core/bpf/allow_bind/allow-bind.c
source from /tmp/tmpk6fxj7j_.o object file
```

[0] https://github.com/libbpf/libbpf/blob/master/src/libbpf.h#L61
[1] https://github.com/libbpf/libbpf/blob/master/src/libbpf.h#L103

Commit:
bpf: add HAVE_LIBBPF, BUILD_BPF config options and feature gate

* Add `build-bpf` feature gate with 'auto', 'true' and 'false' choices
* Add libbpf [0] dependency
* Search for clang and llc binaries the build environment.

For libbpf [0], make 0.2.0 [1] the minimum required version.
If libbpf is satisfied, set HAVE_LIBBPF config option to 1.

If `build-bpf` feature gate is set to 'auto', whether feature is enabled
or disabled is defined by presence of all of libbpf, clang and llvm in build
environment. With 'auto' all dependencies are optional.
If the gate is set to `true`, make all of the libbpf, clang and llvm
dependencies mandatory.
If it's set to `false`, set `BUILD_BPF` to false and make libbpf
dependency optional.

libbpf dep is dynamic followed by the common pattern in systemd.
find_program doesn't allow to set minimum version similary to
`dependency` option. The most recent BPF features include BTF which
require minimim v.10 LLVM, allow_bind program doesn't use BTF features
and builds with clang and llvm 9.0.

Commit:
bpf: add helpers to attach and detach multiple progs

Introduce cgroup_bpf_attach_programs and
cgroup_bpf_detach_programs helpers iterating over a set of cgroup-bpf
progs defined by BPFProgramV2 type.

If libbpf dependency is not satisfied, return -ENOTSUP.

Commit:
readme: update README with requirements for AllowBindPorts=

mauriciovasquezbernal committed:
core: add RestrictNetworkInterfaces= BPF program source code

The code is composed by two BPF_PROG_TYPE_CGROUP_SKB programs that
are loaded in the cgroup inet ingress and egress hooks
(BPF_CGROUP_INET_{INGRESS|EGRESS}).

The decision to let a packet pass or not is based on a map that contains
the ifindexes of the interfaces. The key 0 of the map is used to signal
whether it's an allow or deny-list.

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
core: add RestrictNetworkInterfaces= functions

This commit adds the following functions()
- restrict_network_interfaces_supported: checks if the kernel has the
features needed to support the BPF programs required.
- restrict_network_interfaces_install() loads and attaches the
RestrictNetworkInterfaces BPF programs.

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
core: add load fragment implementation for RestrictNetworkInterfaces=

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
systemctl: show RestrictNetworkInterfaces= in systemctl show

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
core: implement RestrictNetworkInterfaces=

Use the previously introduced functions to load and attach the BPF
programs when a unit is created.

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
core: add D-bus properties for RestrictNetworkInterfaces=

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
Document RestrictNetworkInterfaces dbus properties

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
man: add RestrictNetworkInterfaces= documentation

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
Add unit tests for RestrictNetworkInterfaces=

Add a unit test that creates a set of veth pairs to test tha the
RestrictNetworkInterfaces= actually blocks traffic in given interfaces.

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mauriciovasquezbernal committed:
Add RestrictNetworkInterfaces= to fuzz

Signed-off-by: Mauricio Vásquez <mauricio@kinvolk.io>

mrc0mmand committed:
ci: make the script names a bit more consistent

mrc0mmand committed:
ci: drop an unnecessary sudo

mrc0mmand committed:
ci: use the recommended meson syntax

See: docs/HACKING.md

fbuihuu committed:
sysusers: flush nscd's caches whenever /etc/{passwd,group} are modified

pothos commented:
Regarding your first point: The bpf-firewall code already sets `BPF_F_ALLOW_MULTI` if custom bpf programs are to be attached or if the cgroup is delegated (`flags`, few lines above): https://github.com/systemd/systemd/blob/fee6441601c979165ebcbb35472036439f8dad5f/src/core/bpf-firewall.c#L710

pothos commented on man/systemd.resource-control.xml:
```diff
@@ -711,6 +711,55 @@
+          default action.
+          </para>
+
+          <para>The loopback interface ("lo") is not treated in any special way, you have to configure it explicitely
```
small typo: `explicitly`

mrc0mmand committed:
Merge pull request #18374 from mrc0mmand/assorted-ci-tweaks

Assorted CI tweaks

mauriciovasquezbernal commented:
> Regarding your first point: The bpf-firewall code already sets `BPF_F_ALLOW_MULTI` if custom bpf programs are to be attached or if the cgroup is delegated (`flags`, few lines above): https://github.com/systemd/systemd/blob/fee6441601c979165ebcbb35472036439f8dad5f/src/core/bpf-firewall.c#L710

I see. I think I can use a similar approach, my only concern is that this is not scalable as the bpf-firewall has to be updated each time a new feature using CGROUP_SKB is used. Do you know what's the reason not to enable MULTI if it's supported instead of having the additional `u->type == UNIT_SLICE || unit_cgroup_delegate(u)` [check](https://github.com/systemd/systemd/blob/fee6441601c979165ebcbb35472036439f8dad5f/src/core/bpf-firewall.c#L701)?

Commit:
Add missing pieces to bpf API

This commit is a reduced version of https://github.com/wat-ze-hex/systemd/commit/431d83f8e74c3ca9e75b962cc650ced6fa9a7a51
only containing the changes related to the bpf API.
